tuned_lens/model_surgery.py

```diff
@@ -52,25 +52,40 @@ def assign_key_path(model: T, key_path: str, value: Any) -> Generator[T, None, N
         set_key_path_(model, key_path, old_value)
 
 
-def get_final_layer_norm(
-    model: PreTrainedModel, norm_class: Type[th.nn.Module] = th.nn.LayerNorm
-) -> Optional[th.nn.Module]:
-    """Use heuristics to find the final layer norm in a model, if it exists."""
-    base = model.base_model
-    if decoder := getattr(base, "decoder", None):
-        base = decoder
+def get_final_layer_norm(model: tr.AutoModelForCausalLM):
```

Comment:
I think we should call this `get_final_norm_layer` in preparation for supporting RMSNorm.

Also nit:  `tr.AutoModelForCausalLM` is not actually the superclass of anything, so the right annotation here would be like `tr.PreTrainedModel`. At least, if you want to go with the usual interpretation of type annotations as asserting that `isinstance(model, tr.AutoModelForCausalLM) is True`

Reply:
Also like, you don't have to instantiate any of these models with `AutoModelForCausalLM`, you could instantiate them directly using their dedicated model classes